netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -711,16 +713,10 @@ private void sendResponseHeaders(ChannelHandlerContext ctx, SendResponseHeadersC
     try (TaskCloseable ignore = PerfMark.traceTask("NettyServerHandler.sendResponseHeaders")) {
       PerfMark.attachTag(cmd.stream().tag());
       PerfMark.linkIn(cmd.getLink());
-      // TODO(carl-mastrangelo): remove this check once https://github.com/netty/netty/issues/6296
-      // is fixed.
       int streamId = cmd.stream().id();
       Http2Stream stream = connection().stream(streamId);
-      if (stream == null) {
-        resetStream(ctx, streamId, Http2Error.CANCEL.code(), promise);
```

Comment:
I agree we probably don't want to `resetStream()` if the stream no longer exists. But since you remove the `return`, we will still do the `writeHeaders()` which we probably don't want to. (It looks like Netty will notice and fail the future. So it "works" but no need to trigger that flow)

It seems we want to return early here and sendGrpcFrame().

Reply:
I think the reason I changed `sendResponseHeaders` and `sendGrpcFrame` to always fall through to `encoder().writeData` or `encoder().writeHeaders` is that it cleanly ensures the `promise` argument is always completed somehow. There's an inherent race—the other side might reset the stream one instant after `writeData` or `writeHeaders` is called failing the promise—, so it seems simplest to call into the encoder even if we could know it will fail in some orderings of the race.

Reply:
Oh, the promise. Good catch. Yeah, we will need to make sure to complete it. But we could just call `promise.setSuccess()` (or failure).

The race is a fair view. That would involve _different_ Netty code, but on the surface that seems fair. Looking at the specific Netty code though, it seems to think that is an incorrect usage and fails the promise with IllegalStateException ([writeHeaders](https://github.com/netty/netty/blob/b611fdebdf7787c214732991fb9142662f2583eb/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L204), [writeData](https://github.com/netty/netty/blob/b611fdebdf7787c214732991fb9142662f2583eb/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L135)). We have poor (read: nonexistent) error handling for many writes, but IllegalStateException would likely cause us to shut down the connection. Those failures also allocate an exception each time, so are probably more costly than if data was already buffered and failed all-at-once with a single exception instance. That's important because this is in the cancellation flow and we want to throw away work as quickly as possible.

You had actually convinced me that that was the way to go, and then I realized Netty's specific behavior discourages it.

Reply:
> The race is a fair view. That would involve _different_ Netty code, but on the surface that seems fair. Looking at the specific Netty code though, it seems to think that is an incorrect usage and fails the promise with IllegalStateException ([writeHeaders](https://github.com/netty/netty/blob/b611fdebdf7787c214732991fb9142662f2583eb/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L204), [writeData](https://github.com/netty/netty/blob/b611fdebdf7787c214732991fb9142662f2583eb/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L135)).

> We have poor (read: nonexistent) error handling for many writes, but IllegalStateException would likely cause us to shut down the connection.

So I tried to audit anyone who ever registers interest in `SendGrpcFrameCommand` or `SendResponseHeadersCommand` promises, and I think it's just these two `NettyServerStream.Sink` methods':
https://github.com/grpc/grpc-java/blob/eba699ad1605d46e8bfb77e0dbdd047ba132f941/netty/src/main/java/io/grpc/netty/NettyServerStream.java#L96-L125
So, I think failures of this promise are simply ignored?

> Those failures also allocate an exception each time, so are probably more costly than if data was already buffered and failed all-at-once with a single exception instance. That's important because this is in the cancellation flow and we want to throw away work as quickly as possible.

I can see the concern with the exception allocation (especially since, per _supra_, it's simply dropped)˙. However, this case should be quite uncommon, right? If the client cancels, the deframer will be abruptly closed and no more `SendGrpcFrameCommand`s or `SendResponseHeaderCommand`s will come in. The only window to hit this exception is if the deframer is writing at the exact moment a stream reset is received.

> 
> You had actually convinced me that that was the way to go, and then I realized Netty's specific behavior discourages it.

I will push up a commit that calls `promise.setFailure()` with a statically allocated exception when `stream == null`, though I'm not 100% convinced such extra code is valuable.

Reply:
> So, I think failures of this promise are simply ignored?

Yes. And that's what I was getting at with the "(read: nonexistent)". But we _should_ be looking at it. https://github.com/grpc/grpc-java/issues/10849 for example. So I'd rather not make that harder. And any error handling here will need to look at the type of class to determine whether to kill the call or the connection. IllegalStateException would probably kill the connection.

> However, this case should be quite uncommon, right? If the client cancels, the deframer will be abruptly closed and no more SendGrpcFrameCommands or SendResponseHeaderCommands will come in.

A properly functioning system could have lots of cancellation. And this can happen _per frame_, for which there could be many if there are larger messages. But the real kicker is "sending is controlled by the application." Yes, the _deframer_ is closed, but the _framer_ continues as long as the application continues writing messages. So it depends on the application to notice the cancellation.